experimental/parser/lex.go

```diff
@@ -197,6 +192,77 @@ func lex(ctx token.Context, errs *report.Report) {
 	fuseStrings(l)
 }
 
+// lexPrelude performs various file-prelude checks, such as size and encoding
+// verification. Returns whether lexing should proceed.
+func lexPrelude(l *lexer) bool {
+	// Check that the file isn't too big. We give up immediately if that's
+	// the case.
+	if len(l.Text()) > maxFileSize {
+		l.Errorf("files larger than 2GB (%d bytes) are not supported", maxFileSize).Apply(
+			report.InFile(l.Path()),
+		)
+		return false
+	}
+
+	// Heuristically check for a UTF-16-encoded file. There are two good
+	// heuristics:
+	// 1. Presence of a UTF-16 BOM, which is either FE FF or FF FE, depending on
+	//    endianness.
+	// 2. Exactly one of the first two bytes is a NUL. Valid Protobuf cannot
+	//    contain a NUL in the first two bytes, so this is probably a UTF-16-encoded
+	//    ASCII rune.
+	bom16 := strings.HasPrefix(l.Text(), "\xfe\xff") || strings.HasPrefix(l.Text(), "\xff\xfe")
+	ascii16 := len(l.Text()) >= 2 && (l.Text()[0] == 0 || l.Text()[1] == 0)
+	if bom16 || ascii16 {
+		l.Errorf("input appears to be encoded with UTF-16").Apply(
+			report.InFile(l.Path()),
+			report.Notef("Protobuf files must be UTF-8 encoded"),
+		)
+		return false
+	}
+
+	// Check that the text of the file is actually UTF-8.
+	var idx int
+	var count int
+	stringsx.Runes(l.Text())(func(n int, r rune) bool {
+		if r == -1 {
+			if count == 0 {
+				idx = n
+			}
+			count++
+		}
+		return true
+	})
+	frac := float64(count) / float64(len(l.Text()))
+	switch {
+	case frac == 0:
+		break
```

Comment:
Super tiny nit: I'm indifferent about changing this, if we wanted to save calculating the fraction, we check up-front `count > 0`

Reply:
Nah, the cost of examining every byte of the input is way higher than the cost of two i2f instructions followed by an fdiv. In a tight loop it would matter, but not here. :P